### ISSUE TITLE
Upgraded to .NET9 - Extension Methods of IQueryable<T> not translated when used as a sub/child query

### ISSUE THREAD

**Issue by PTAdvanced** [dotnet/efcore#35114]
## File a bug

I have just migrated/updated a working code base from .NET8 to .NET9.

Once complete, I find that EF LINQ Queries which are utilising Extensions of IQueryable<T> in a child/sub query are not translated and resulting in an exception.

I have looked through the release notes and looked for any information on breaking changes on how EF deals with evaluating expressions, or child expressions, but I cannot find anything mentioned.

### Include your code

This is much the same issue as described in this old issue from EF6 https://github.com/dotnet/ef6/issues/98

In my code base, the exception encountered is:

The LINQ expression 'DbSet<PointState>()
    .ExcludeDeleted()
    .ExcludePointsNotInUse()
    .Join(
        inner: DbSet<ZoneState>(), 
        outerKeySelector: point => new { 
            SiteId = point.SiteId, 
            Number = point.Zone.Number, 
            SystemId = point.SystemId
         }, 
        innerKeySelector: zone => new { 
            SiteId = zone.SiteId, 
            Number = zone.ZoneNumber, 
            SystemId = zone.SystemId
         }, 
        resultSelector: (point, zone) => new { 
            point = point, 
            zone = zone
         })' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'. See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.

In this case, the query shown here is actually a subquery, specifically used in a Join as the inner expression.

A simplified version of this query would be:

var myResults = _myContext.Customers
    .Join(
        _myContext.Points
            .ExcludeDeleted()
            .ExcludePointsNotInUse(),
             customer => customerId,
             point => point.customerId,
             (customer, point) => new {customer, point} );

Where .ExcludeDeleted() is as follows:

    internal static IQueryable<PointState> ExcludeDeleted(this IQueryable<PointState> queryable)
    {
        return queryable.Where(p => p.DeletionState.State != DigitalTwinItemDeletionState.Deleted.State);
    }

I can refactor my code as follows (and as suggested in the old EF6 link https://github.com/dotnet/ef6/issues/98) by doing the following:

var pointsInUse = _myContext.Points
            .ExcludeDeleted()
            .ExcludePointsNotInUse();

var myResults = _myContext.Customers
    .Join(
        pointsInUse,
             customer => customerId,
             point => point.customerId,
             (customer, point) => new {customer, point});

Essentially splitting the query composition.

So I believe this proves that .NET9 seems to have regressed to the child expression evaluation found in EF6.

As I say, the above code with the child expression utilising the extension methods worked in .NET 5, .NET6, .NET7 and .NET8. The behaviour has only now changed when upgrading to .NET9.

**Comment by roji** [comment]
Can you submit a minimal code sample? It's difficult to see from the various disconnected snippets above exactly what you're doing. In principle, top-level queryable methods aren't even seen by EF - they're executed as regular methods in .NET, and simply compose whatever LINQ operators they compose (e.g. Where in your ExecuteDelete example above). So there really shouldn't be any issues with such methods.

**Comment by PTAdvanced** [comment]
Yes, I will try and pull something together quickly now.

**Comment by PTAdvanced** [comment]
I have tried to recreate the issue here 

https://github.com/PTAdvanced/NET9EFSubQueryIssue/tree/main 

and I have a branch there which is built on .NET9.

However, in this very simple example it's working as expected.

I need to try and understand what the scenario is in our more complex code base.

I have confirmed though that rolling back to .NET8, our code works. The same code on .NET9 fails. And the solution to our .NET issue is to decompose the query into 2 parts, i.e. assign the inner query to a local variable, and then reference the variable in the main query.

If I can get a simple version of the issue, I will update here
